src/NHibernate/Linq/UpdateSyntax.cs

```diff
@@ -0,0 +1,65 @@
+using System;
+using System.Linq.Expressions;
+
+namespace NHibernate.Linq
+{
+	/// <summary>
+	/// An update object on which values to update can be specified.
+	/// </summary>
+	/// <typeparam name="T">The type of the entities to update.</typeparam>
+	public class UpdateSyntax<T>
+	{
+		private readonly Expression _sourceExpression;
+		private readonly INhQueryProvider _provider;
+		private readonly bool _versioned;
+
+		internal UpdateSyntax(Expression sourceExpression, INhQueryProvider provider, bool versioned)
+		{
+			_sourceExpression = sourceExpression;
+			_provider = provider;
+			_versioned = versioned;
+		}
+
+		/// <summary>
+		/// Specify the assignments and execute the update.
+		/// </summary>
+		/// <param name="assignmentActions">The assignments.</param>
+		/// <returns>The number of updated entities.</returns>
+		public int Assign(Action<Assignments<T, T>> assignmentActions)
+		{
+			if (assignmentActions == null)
+				throw new ArgumentNullException(nameof(assignmentActions));
+			var assignments = new Assignments<T, T>();
+			assignmentActions.Invoke(assignments);
+
+			return ExecuteUpdate(DmlExpressionRewriter.PrepareExpression<T>(_sourceExpression, assignments.List));
+		}
+
+		/// <summary>
+		/// Specify the assignments and execute the update.
+		/// </summary>
+		/// <param name="expression">The assignments expressed as a member initialization, e.g.
+		/// <c>x => new Dog { Name = x.Name, Age = x.Age + 5 }</c>. Unset members are ignored and left untouched.</param>
+		/// <returns>The number of updated entities.</returns>
+		public int As(Expression<Func<T, T>> expression)
```

Comment:
Yes, as anticipated, and handled by this [check](https://github.com/nhibernate/nhibernate-core/pull/391/commits/63d659e6178420e132e505d4cc1689109939561e#diff-db3b5a31ee1bd41be64a8166f7d27707R160), added in ["Protect against invalid use cases" commit](https://github.com/nhibernate/nhibernate-core/pull/391/commits/6aa9c1807716958e59739b612e7feb38db188f08).

This invalid usage case correctly errors out, telling this strongly typed method expect the exact same type.

If doing `session.Query<Car>().Update().As(c=>new Animal(){})`, it will go through the anonymous object version and there the anonymous object method will complains it accepts only anonymous objects.

If we merge both, it will complicate the checking and the error message. Where we do the checking, we do not know if it is called for an update or an insert, so we may end-up with some cumbersome message trying to cover all cases. (Strongly typed update? Accept only exact same type, but need to not imply that if raising the error for an insert. Anonymous insert? Need an explicitly specified target type, while update do not care.)
Otherwise we would have to move up the checking, meaning starting to analyze the expression in `InsertSyntax` and `UpdateSyntax`, instead of keeping it together in `DmlExpressionRewriter`. (Or tell to `DmlExpressionRewriter` if it is called for update or insert in order to adapt the error message.)

I personally do not feel that having both methods is unneeded. It eases implementation and understanding of this API. The later alone seems enough to me for having both methods.